frontend/src/metabase/query_builder/containers/QueryBuilder.jsx

```diff
@@ -354,19 +364,62 @@ function QueryBuilder(props) {
     }
   });
 
+  const { isRunning } = uiControls;
+
+  const [sendNotification, setSendNotification] = useState(false);
+  const [showToaster, setShowToaster] = useState(false);
+
+  const toastTrigger = useLoadingTimer(isRunning, 5000);
+  const [requestPermission, showNotification] = useWebNotification();
+
+  useEffect(() => {
+    if (toastTrigger) {
+      setShowToaster(true);
+    }
+  }, [toastTrigger]);
```

Comment:
WDYT about passing a callback to `useLoadingTimer` to avoid `useEffect` here? For example:

```jsx
useLoadingTimer(isRunning, {
   timeout: 5000,
   onTimeout: () => setShowToaster(true)
})
```